File: packages/ipc/src/asset.ts
import { type AssetFormat, type AssetUploadResult, commands } from '@poietica/contract'
import { throughIpc } from './error'

/*
 * 资产会话：一批图片挂在一个令牌下面，关掉就一起释放。
 *
 * 这一层只做一件事 —— 把原生那三条命令包成"一次调用只有一条路"的形状（见
 * error.ts 的 throughIpc）。形状不在这里重新定义：AssetUploadResult 直接取自
 * 生成绑定，Rust 侧的类型是权威。
 *
 * 字节不出现在这个文件里，也不出现在任何一条命令的参数里。拖放与文件对话框
 * 交给渲染层的是路径，读盘发生在原生侧：webview 不该为了把一张 4 MB 的 PNG
 * 交给本机进程，先把它读进内存、编码、再送回去。
 */

/** 原生按路径入库之后交回来的那一份。source 就是 <img src> 能直接用的地址。 */
export type AssetImport = AssetUploadResult

/** 一种收得下的格式：内容类型，加上它在系统对话框里的扩展名。 */
export type { AssetFormat }

/**
 * 原生收得下的格式清单。
 *
 * 扩展名只给系统对话框的过滤器用，不是判据 —— 判据是文件头，在原生那一侧，
 * 而且两者出自同一张表（commands/asset.rs 的 FORMATS）。这一层因此不持有
 * 任何格式知识，它只是把那张表运过来。
 */
export function listAssetFormats(): Promise<readonly AssetFormat[]> {
  return throughIpc(() => commands.assetFormats())
}

/** 开一条资产会话，拿到它的令牌。 */
export function openAssetSession(): Promise<string> {
  return throughIpc(async () => {
    const opened = await commands.assetSessionOpen()

    return opened.sessionToken
  })
}

/**
 * 把这些路径读进会话，顺序与传入一致。
 *
 * 内容类型由原生按文件头判定，不看扩展名，也不看渲染层的猜测；认不出来的格式
 * 整批拒绝，所以调用方拿到的每一项都是能投递的。
 */
export function importAssets(
  sessionToken: string,
  paths: readonly string[],
): Promise<readonly AssetImport[]> {
  /* readonly 的数组与生成绑定要的可变数组是两个类型，所以复制一次 ——
  数组复制只在这一层做。 */
  return throughIpc(() => commands.assetImport({ sessionToken, paths: [...paths] }))
}

/**
 * 剪贴板里的那一张图。
 *
 * 三条进门的路里只有这一条要经过字节：截图是一团没有名字也没有路径的 blob，
 * 系统给不出路径，所以它走不了 importAssets。拖放与文件对话框交的都是路径。
 *
 * 内容类型不在参数里。它由原生按文件头判定，与 importAssets 共用同一个判据 ——
 * 渲染层报的 `File.type` 来自扩展名，而资产协议是带 nosniff 投递的。
 */
export function uploadAsset(sessionToken: string, base64: string): Promise<AssetImport> {
  return throughIpc(() => commands.assetUpload({ sessionToken, base64 }))
}

/** 从会话里放掉一张。输入框里被移除的那一张不该继续占着注册表的预算。 */
export function removeAsset(sessionToken: string, assetToken: string): Promise<void> {
  return throughIpc(async () => {
    await commands.assetRemove({ sessionToken, assetToken })
  })
}

File: apps/desktop/src-tauri/src/asset.rs
//! 资产的 IPC 编码、执行器选择与脱敏错误。

use base64::Engine as _;
use base64::engine::general_purpose::STANDARD as BASE64;
use serde::{Deserialize, Serialize};
use specta::Type;
use tauri::{AppHandle, State, async_runtime};
use uuid::Uuid;

use crate::error::Error;
use crate::paths;
use poietica_asset::{
    AssetIntakeError, AssetProtocolError, AssetProtocolRegistry, ImportedAsset, ImportedKind,
    MAX_ASSET_BYTES, import_bytes, import_files,
};
use poietica_problem::Problem;

type CommandResult<T> = Result<T, Problem>;

#[derive(Clone, Debug, Deserialize, Type)]
#[serde(rename_all = "camelCase")]
pub struct AssetUploadRequest {
    pub session_token: String,
    /// base64 原始字节，不带 `data:` 前缀；刻意不用 `Vec<u8>`：JSON IPC 下它线上是 `number[]`，大四五倍（见 Tauri v2 InvokeArgs）。
    pub base64: String,
}

#[derive(Clone, Debug, Deserialize, Type)]
#[serde(rename_all = "camelCase")]
pub struct AssetImportRequest {
    pub session_token: String,
    pub paths: Vec<String>,
}

#[derive(Clone, Debug, Serialize, Type)]
#[serde(rename_all = "camelCase")]
pub struct AssetSessionResult {
    pub session_token: String,
}

#[derive(Clone, Copy, Debug, Deserialize, Eq, PartialEq, Serialize, Type)]
#[serde(rename_all = "camelCase")]
pub enum AssetKind {
    Image,
    File,
}

impl From<ImportedKind> for AssetKind {
    fn from(kind: ImportedKind) -> Self {
        match kind {
            ImportedKind::Image => Self::Image,
            ImportedKind::File => Self::File,
        }
    }
}

#[derive(Clone, Debug, Serialize, Type)]
#[serde(rename_all = "camelCase")]
pub struct AssetUploadResult {
    /// Image：进内存注册表、source 是预览地址；File：落盘暂存、source 为空。
    pub kind: AssetKind,
    pub asset_token: String,
    pub content_hash: String,
    pub source: String,
    pub byte_length: u32,
    pub content_type: String,
}

#[derive(Clone, Debug, Deserialize, Type)]
#[serde(rename_all = "camelCase")]
pub struct AssetRemoveRequest {
    pub session_token: String,
    pub asset_token: String,
}

#[derive(Clone, Debug, Deserialize, Type)]
#[serde(rename_all = "camelCase")]
pub struct AssetSessionCloseRequest {
    pub session_token: String,
}

/// 调用方只见脱敏后的 IPC 文案，永远拿不到原生细节。
#[tauri::command]
#[specta::specta]
pub async fn asset_session_open(
    assets: State<'_, AssetProtocolRegistry>,
) -> CommandResult<AssetSessionResult> {
    let session_token = Uuid::now_v7().simple().to_string();

    assets
        .open_session(&session_token)
        .map_err(map_asset_error)?;

    Ok(AssetSessionResult { session_token })
}

#[tauri::command]
#[specta::specta]
pub async fn asset_upload(
    request: AssetUploadRequest,
    assets: State<'_, AssetProtocolRegistry>,
) -> CommandResult<AssetUploadResult> {
    let registry = assets.inner().clone();
    async_runtime::spawn_blocking(move || {
        if request.base64.len() > MAX_ASSET_BYTES.div_ceil(3) * 4 {
            return Err(map_asset_error(AssetProtocolError::AssetTooLarge));
        }
        let bytes = BASE64.decode(request.base64.as_bytes()).map_err(|_| {
            Problem::from(Error::Validation("attachment is not valid base64".into()))
        })?;
        import_bytes(&registry, &request.session_token, bytes)
            .map(AssetUploadResult::from)
            .map_err(map_intake_error)
    })
    .await
    .map_err(|cause| {
        log::warn!("asset ingestion task failed: {cause}");
        Problem::from(Error::Internal("asset ingestion task failed".into()))
    })?
}

#[tauri::command]
#[specta::specta]
pub async fn asset_import(
    app: AppHandle,
    request: AssetImportRequest,
    assets: State<'_, AssetProtocolRegistry>,
) -> CommandResult<Vec<AssetUploadResult>> {
    let registry = assets.inner().clone();
    let staging = paths::composer_staging_root(&app).map_err(Problem::from)?;
    async_runtime::spawn_blocking(move || {
        import_files(&registry, &request.session_token, &staging, &request.paths)
            .map(|items| items.into_iter().map(AssetUploadResult::from).collect())
            .map_err(map_intake_error)
    })
    .await
    .map_err(|cause| {
        log::warn!("asset ingestion task failed: {cause}");
        Problem::from(Error::Internal("asset ingestion task failed".into()))
    })?
}

impl From<ImportedAsset> for AssetUploadResult {
    fn from(asset: ImportedAsset) -> Self {
        Self {
            kind: asset.kind.into(),
            asset_token: asset.content_hash.clone(),
            content_hash: asset.content_hash,
            source: asset.source,
            byte_length: asset.byte_length,
            content_type: asset.content_type,
        }
    }
}

fn map_intake_error(error: AssetIntakeError) -> Problem {
    log::warn!("asset ingestion failed: {error}");
    match error {
        AssetIntakeError::Protocol(cause) => map_asset_error(cause),
        AssetIntakeError::Blob(cause) => {
            log::error!("an attachment could not be staged: {cause}");
            Error::Asset("an attachment could not be stored".into()).into()
        }
        AssetIntakeError::Read(_) => Error::NotFound("file could not be read".into()).into(),
    }
}

#[tauri::command]
#[specta::specta]
pub async fn asset_remove(
    request: AssetRemoveRequest,
    assets: State<'_, AssetProtocolRegistry>,
) -> CommandResult<()> {
    let removed = assets
        .remove(&request.session_token, &request.asset_token)
        .map_err(map_asset_error)?;

    // 通用文件不进内存注册表（在 tmp 暂存，启动对账清）：查无此项不是错误。
    // 记 warn 而不是 debug：这一支同样接得住拼错的图片令牌，静默会把真错误埋掉。
    if !removed {
        log::warn!("asset {} is not held by the registry", request.asset_token);
    }

    Ok(())
}

#[tauri::command]
#[specta::specta]
pub async fn asset_session_close(
    request: AssetSessionCloseRequest,
    assets: State<'_, AssetProtocolRegistry>,
) -> CommandResult<()> {
    assets
        .remove_session(&request.session_token)
        .map_err(map_asset_error)?;

    Ok(())
}

fn map_asset_error(error: AssetProtocolError) -> Problem {
    let error = match error {
        AssetProtocolError::InvalidToken
        | AssetProtocolError::InvalidContentHash
        | AssetProtocolError::UnsupportedContentType
        | AssetProtocolError::AssetTooLarge => Error::Validation("invalid asset request".into()),

        AssetProtocolError::NotFound => Error::NotFound("asset session or asset not found".into()),

        AssetProtocolError::RegistryBudgetExceeded
        | AssetProtocolError::DuplicateAsset
        | AssetProtocolError::ReferenceOverflow => {
            Error::Asset("asset registry rejected resource".into())
        }

        AssetProtocolError::Internal => Error::Internal("asset registry unavailable".into()),
    };

    error.into()
}

#[cfg(test)]
mod tests {
    #![allow(
        clippy::expect_used,
        clippy::unwrap_used,
        clippy::panic,
        clippy::indexing_slicing,
        clippy::shadow_unrelated,
        reason = "tests operate on known-good fixtures; a broken assumption must fail the test loudly"
    )]

    use super::{AssetProtocolError, map_asset_error};
    use poietica_asset::sniff;
    use poietica_problem::Code;
    use sha2::{Digest, Sha256};

    #[test]
    fn content_hash_is_canonical_sha256() {
        let hash = hex::encode(Sha256::digest(b"asset"));

        assert_eq!(hash.len(), 64);
        assert!(
            hash.bytes()
                .all(|byte| { byte.is_ascii_digit() || matches!(byte, b'a'..=b'f') })
        );
    }

    #[test]
    fn content_type_comes_from_the_bytes_not_the_name() {
        assert_eq!(
            sniff(b"\x89PNG\r\n\x1a\n\x00\x00\x00\rIHDR"),
            Some("image/png")
        );
        assert_eq!(sniff(&[0xFF, 0xD8, 0xFF, 0xE0]), Some("image/jpeg"));
        assert_eq!(sniff(b"RIFF\x00\x00\x00\x00WEBPVP8 "), Some("image/webp"));

        assert_eq!(
            sniff(b"<svg xmlns=\"http://www.w3.org/2000/svg\"/>"),
            Some("text/plain")
        );
        assert_eq!(sniff(b"plain text"), Some("text/plain"));
        assert_eq!(sniff(b""), None);
    }

    #[test]
    fn asset_errors_do_not_expose_internal_details() {
        let problem = map_asset_error(AssetProtocolError::RegistryBudgetExceeded);

        assert_eq!(problem.code, Code::AssetRejected);
        assert!(problem.details.is_empty());
    }
}
